single-validator-manager/program/src/instruction.rs

```diff
@@ -0,0 +1,694 @@
+    ///  10. `[]` Stake program id
+    ///  11. `[]` Token program id
+    ///  12. `[]` Stake pool program id
+    CreateStakePool,
+
+    ///   Adds the single validator stake account to the pool.
+    ///
+    ///   The stake account will have the rent-exempt amount plus
+    ///   `max(crate::MINIMUM_ACTIVE_STAKE, solana_program::stake::tools::get_minimum_delegation())`.
+    ///   It is funded from the stake pool reserve.
+    ///
+    ///   0. `[w]` Stake pool
+    ///   1. `[]` Staker
+    ///   2. `[w]` Reserve stake account
+    ///   3. `[]` Stake pool withdraw authority
+    ///   4. `[w]` Validator stake list storage account
+    ///   5. `[w]` Stake account to add to the pool
+    ///   6. `[]` Validator vote account for the single validator stake pool
+    ///   7. `[]` Rent sysvar
+    ///   8. `[]` Clock sysvar
+    ///   9. '[]' Stake history sysvar
+    ///  10. '[]' Stake config sysvar
+    ///  11. `[]` System program
+    ///  12. `[]` Stake program
+    ///  13. `[]` Stake pool program
+    ///
+    ///  userdata: optional non-zero u32 seed used for generating the validator
+    ///  stake address
+    AddValidatorToPool,
+
+    ///   Removes a non-canonical validator from the pool, deactivating its stake
+    ///
+    ///   This is useful if an existing stake pool is assigned to the management
+    ///   program, but has additional validators.
+    ///
+    ///   0. `[]` Validator vote account for the single validator stake pool
+    ///   1. `[w]` Stake pool
+    ///   2. `[]` Staker
+    ///   3. `[]` Stake pool withdraw authority
+    ///   4. `[w]` Validator stake list storage account
+    ///   5. `[w]` Stake account to remove from the pool
+    ///   6. `[]` Transient stake account, to check that that we're not trying to activate
+    ///   7. `[]` Sysvar clock
+    ///   8. `[]` Stake program
+    ///   9. `[]` Stake pool program
+    RemoveValidatorFromPool,
+
+    /// Decrease active stake on a validator, eventually moving it to the reserve
+    ///
+    /// The instruction only succeeds if the stake pool reserve has less than
+    /// the minimum delegation amount. It allows small pool token holders to
+    /// withdraw their tokens from the pool.
+    ///
+    /// This instruction can also be used on additional validators that are not
+    /// the single validator, if an existing stake pool was reassigned to this
+    /// management program.
+    ///
+    ///  0. `[]` Validator vote account for the single validator stake pool
+    ///  1. `[]` Stake pool
+    ///  2. `[]` Stake pool staker
+    ///  3. `[]` Stake pool withdraw authority
+    ///  4. `[w]` Validator list
+    ///  5. `[w]` Canonical stake account to split from
+    ///  6. `[w]` Transient stake account to receive split
+    ///  7. `[]` Clock sysvar
+    ///  8. `[]` Rent sysvar
+    ///  9. `[]` System program
+    /// 10. `[]` Stake program
+    /// 11. `[]` Stake pool program
+    ///
+    /// userdata: seed to use for transient stake account
```

Comment:
Yeah that's true! There are quite a few annoying movements of stake accounts

> we should also allow splitting directly into a mergeable destination sweat_smile

I've thought about this one a lot, but since we have to change ownership, it wouldn't actually benefit stake pools, unless splitting into a "mergeable" destination doesn't look at the "authorized" portion.

Reply:
> it wouldn't actually benefit stake pools, unless splitting into a "mergeable" destination doesn't look at the "authorized" portion.

yeah true.  definitely wouldn't want to skip the `authorized` check i don't think.  my mind was on the SOL deposit/withdraw process when this occurred to me